File: backend/run_security_tests.py
#!/usr/bin/env python3
"""
Comprehensive security and privacy testing runner.
This script runs all security-related tests and generates a security report.
"""

import subprocess
import sys
import json
import time
from datetime import datetime
from pathlib import Path
import argparse


class SecurityTestRunner:
    """Run comprehensive security and privacy tests."""

    def __init__(self):
        self.results = {
            "test_run_info": {
                "timestamp": datetime.utcnow().isoformat(),
                "python_version": sys.version,
                "test_categories": []
            },
            "test_results": {},
            "summary": {
                "total_tests": 0,
                "passed": 0,
                "failed": 0,
                "skipped": 0,
                "errors": 0
            },
            "security_issues": [],
            "recommendations": []
        }

    def run_test_category(self, category_name, test_file, description):
        """Run a specific category of security tests."""

        print(f"\n{'='*60}")
        print(f"Running {category_name}")
        print(f"Description: {description}")
        print(f"Test file: {test_file}")
        print(f"{'='*60}")

        start_time = time.time()

        try:
            # Run pytest with verbose output and JSON report
            cmd = [
                "python", "-m", "pytest",
                test_file,
                "-v",
                "--tb=short",
                "--json-report",
                f"--json-report-file=test_reports/{category_name}_report.json"
            ]

            result = subprocess.run(
                cmd,
                capture_output=True,
                text=True,
                cwd=Path(__file__).parent
            )

            end_time = time.time()
            duration = end_time - start_time

            # Parse results
            test_result = {
                "category": category_name,
                "description": description,
                "duration": duration,
                "return_code": result.returncode,
                "stdout": result.stdout,
                "stderr": result.stderr,
                "status": "PASSED" if result.returncode == 0 else "FAILED"
            }

            # Try to load JSON report if available
            json_report_path = Path(
                f"test_reports/{category_name}_report.json")
            if json_report_path.exists():
                try:
                    with open(json_report_path, 'r') as f:
                        json_report = json.load(f)
                        test_result["detailed_results"] = json_report

                        # Update summary
                        if "summary" in json_report:
                            summary = json_report["summary"]
                            self.results["summary"]["total_tests"] += summary.get(
                                "total", 0)
                            self.results["summary"]["passed"] += summary.get(
                                "passed", 0)
                            self.results["summary"]["failed"] += summary.get(
                                "failed", 0)
                            self.results["summary"]["skipped"] += summary.get(
                                "skipped", 0)
                            self.results["summary"]["errors"] += summary.get(
                                "error", 0)

                except Exception as e:
                    print(f"Warning: Could not parse JSON report: {e}")

            self.results["test_results"][category_name] = test_result
            self.results["test_run_info"]["test_categories"].append(
                category_name)

            # Print summary
            if result.returncode == 0:
                print(f"✅ {category_name} PASSED ({duration:.2f}s)")
            else:
                print(f"❌ {category_name} FAILED ({duration:.2f}s)")
                print(f"Error output: {result.stderr}")

            return result.returncode == 0

        except Exception as e:
            print(f"❌ Error running {category_name}: {e}")

            test_result = {
                "category": category_name,
                "description": description,
                "duration": 0,
                "return_code": -1,
                "error": str(e),
                "status": "ERROR"
            }

            self.results["test_results"][category_name] = test_result
            return False

    def analyze_security_issues(self):
        """Analyze test results for security issues."""

        security_issues = []
        recommendations = []

        for category, result in self.results["test_results"].items():
            if result["status"] == "FAILED":
                # Analyze failed tests for security implications

                if "authentication" in category.lower():
                    security_issues.append({
                        "category": "Authentication",
                        "severity": "HIGH",
                        "issue": f"Authentication security tests failed in {category}",
                        "description": "Authentication vulnerabilities detected"
                    })
                    recommendations.append({
                        "category": "Authentication",
                        "priority": "HIGH",
                        "recommendation": "Review and fix authentication security issues immediately"
                    })

                elif "privacy" in category.lower():
                    security_issues.append({
                        "category": "Privacy",
                        "severity": "HIGH",
                        "issue": f"Privacy compliance tests failed in {category}",
                        "description": "Privacy compliance violations detected"
                    })
                    recommendations.append({
                        "category": "Privacy",
                        "priority": "HIGH",
                        "recommendation": "Address privacy compliance issues before deployment"
                    })

                elif "penetration" in category.lower():
                    security_issues.append({
                        "category": "Penetration Testing",
                        "severity": "CRITICAL",
                        "issue": f"Penetration tests failed in {category}",
                        "description": "Security vulnerabilities detected through penetration testing"
                    })
                    recommendations.append({
                        "category": "Security",
                        "priority": "CRITICAL",
                        "recommendation": "Fix security vulnerabilities immediately before any deployment"
                    })

                elif "encryption" in category.lower():
                    security_issues.append({
                        "category": "Encryption",
                        "severity": "HIGH",
                        "issue": f"Encryption security tests failed in {category}",
                        "description": "Data encryption and storage security issues detected"
                    })
                    recommendations.append({
                        "category": "Encryption",
                        "priority": "HIGH",
                        "recommendation": "Review and strengthen data encryption mechanisms"
                    })

        # Add general recommendations
        if self.results["summary"]["failed"] > 0:
            recommendations.append({
                "category": "General",
                "priority": "HIGH",
                "recommendation": "Conduct security code review and fix all failing security tests"
            })

        if self.results["summary"]["total_tests"] == 0:
            security_issues.append({
                "category": "Testing",
                "severity": "MEDIUM",
                "issue": "No security tests were executed",
                "description": "Security testing coverage is insufficient"
            })
            recommendations.append({
                "category": "Testing",
                "priority": "MEDIUM",
                "recommendation": "Implement comprehensive security testing suite"
            })

        self.results["security_issues"] = security_issues
        self.results["recommendations"] = recommendations

    def generate_report(self, output_file="security_test_report.json"):
        """Generate comprehensive security test report."""

        # Analyze results
        self.analyze_security_issues()

        # Calculate additional metrics
        total_tests = self.results["summary"]["total_tests"]
        if total_tests > 0:
            pass_rate = (self.results["summary"]["passed"] / total_tests) * 100
            self.results["summary"]["pass_rate"] = round(pass_rate, 2)
        else:
            self.results["summary"]["pass_rate"] = 0

        # Security score (simple calculation)
        if total_tests > 0:
            security_score = max(
                0, 100 - (len(self.results["security_issues"]) * 10))
            self.results["summary"]["security_score"] = security_score
        else:
            self.results["summary"]["security_score"] = 0

        # Write report
        with open(output_file, 'w') as f:
            json.dump(self.results, f, indent=2)

        print(f"\n📊 Security test report generated: {output_file}")

        return self.results

    def print_summary(self):
        """Print test summary to console."""

        print(f"\n{'='*60}")
        print("SECURITY TEST SUMMARY")
        print(f"{'='*60}")

        summary = self.results["summary"]

        print(f"Total Tests: {summary['total_tests']}")
        print(f"Passed: {summary['passed']} ✅")
        print(f"Failed: {summary['failed']} ❌")
        print(f"Skipped: {summary['skipped']} ⏭️")
        print(f"Errors: {summary['errors']} 💥")
        print(f"Pass Rate: {summary.get('pass_rate', 0):.1f}%")
        print(f"Security Score: {summary.get('security_score', 0)}/100")

        # Print security issues
        if self.results["security_issues"]:
            print(
                f"\n🚨 SECURITY ISSUES DETECTED ({len(self.results['security_issues'])})")
            for issue in self.results["security_issues"]:
                severity_emoji = {
                    "CRITICAL": "🔴",
                    "HIGH": "🟠",
                    "MEDIUM": "🟡",
                    "LOW": "🟢"
                }.get(issue["severity"], "⚪")

                print(
                    f"  {severity_emoji} {issue['severity']}: {issue['issue']}")

        # Print recommendations
        if self.results["recommendations"]:
            print(
                f"\n💡 RECOMMENDATIONS ({len(self.results['recommendations'])})")
            for rec in self.results["recommendations"]:
                priority_emoji = {
                    "CRITICAL": "🔴",
                    "HIGH": "🟠",
                    "MEDIUM": "🟡",
                    "LOW": "🟢"
                }.get(rec["priority"], "⚪")

                print(
                    f"  {priority_emoji} {rec['priority']}: {rec['recommendation']}")

        print(f"\n{'='*60}")


def main():
    """Main function to run security tests."""

    parser = argparse.ArgumentParser(
        description="Run comprehensive security and privacy tests")
    parser.add_argument("--category", help="Run specific test category only")
    parser.add_argument(
        "--output", default="security_test_report.json", help="Output report file")
    parser.add_argument("--no-report", action="store_true",
                        help="Skip generating report file")

    args = parser.parse_args()

    # Create test reports directory
    Path("test_reports").mkdir(exist_ok=True)

    runner = SecurityTestRunner()

    # Define test categories
    test_categories = [
        {
            "name": "authentication_security",
            "file": "tests/test_security_authentication.py",
            "description": "Authentication and authorization security tests"
        },
        {
            "name": "privacy_compliance",
            "file": "tests/test_privacy_compliance.py",
            "description": "Privacy compliance and data protection tests"
        },
        {
            "name": "penetration_testing",
            "file": "tests/test_penetration_testing.py",
            "description": "Penetration testing for security vulnerabilities"
        },
        {
            "name": "data_encryption_security",
            "file": "tests/test_data_encryption_security.py",
            "description": "Data encryption and storage security tests"
        },
        {
            "name": "comprehensive_e2e",
            "file": "tests/test_comprehensive_e2e.py",
            "description": "Comprehensive end-to-end security workflows"
        },
        {
            "name": "performance_load",
            "file": "tests/test_performance_load.py",
            "description": "Performance and load testing for security implications"
        },
        {
            "name": "api_integration_comprehensive",
            "file": "tests/test_api_integration_comprehensive.py",
            "description": "Comprehensive API integration security tests"
        }
    ]

    print("🔒 Starting Comprehensive Security and Privacy Testing")
    print(f"Timestamp: {datetime.utcnow().isoformat()}")

    # Run specific category or all categories
    if args.category:
        # Find and run specific category
        category = next(
            (cat for cat in test_categories if cat["name"] == args.category), None)
        if category:
            runner.run_test_category(
                category["name"], category["file"], category["description"])
        else:
            print(f"❌ Category '{args.category}' not found")
            print("Available categories:")
            for cat in test_categories:
                print(f"  - {cat['name']}: {cat['description']}")
            return 1
    else:
        # Run all categories
        for category in test_categories:
            runner.run_test_category(
                category["name"], category["file"], category["description"])

    # Generate report
    if not args.no_report:
        runner.generate_report(args.output)

    # Print summary
    runner.print_summary()

    # Return appropriate exit code
    if runner.results["summary"]["failed"] > 0 or runner.results["security_issues"]:
        print("\n❌ Security tests failed or security issues detected!")
        return 1
    else:
        print("\n✅ All security tests passed!")
        return 0


if __name__ == "__main__":
    sys.exit(main())

File: mobile/run_security_tests.js
#!/usr/bin/env node

/**
 * Mobile app security testing runner
 * Runs security-focused tests for the React Native application
 */

const { execSync, spawn } = require('child_process');
const fs = require('fs');
const path = require('path');

class MobileSecurityTestRunner {
    constructor() {
        this.results = {
            testRunInfo: {
                timestamp: new Date().toISOString(),
                nodeVersion: process.version,
                testCategories: []
            },
            testResults: {},
            summary: {
                totalTests: 0,
                passed: 0,
                failed: 0,
                skipped: 0
            },
            securityIssues: [],
            recommendations: []
        };
    }

    async runTestCategory(categoryName, testPattern, description) {
        console.log(`\n${'='.repeat(60)}`);
        console.log(`Running ${categoryName}`);
        console.log(`Description: ${description}`);
        console.log(`Test pattern: ${testPattern}`);
        console.log(`${'='.repeat(60)}`);

        const startTime = Date.now();

        try {
            // Run Jest with specific test pattern
            const jestCmd = [
                'npx', 'jest',
                testPattern,
                '--verbose',
                '--json',
                '--outputFile', `test-reports/${categoryName}-report.json`,
                '--coverage',
                '--coverageDirectory', `test-reports/${categoryName}-coverage`
            ];

            const result = execSync(jestCmd.join(' '), {
                encoding: 'utf8',
                stdio: 'pipe'
            });

            const endTime = Date.now();
            const duration = (endTime - startTime) / 1000;

            const testResult = {
                category: categoryName,
                description: description,
                duration: duration,
                status: 'PASSED',
                output: result
            };

            // Try to load JSON report
            const reportPath = path.join('test-reports', `${categoryName}-report.json`);
            if (fs.existsSync(reportPath)) {
                try {
                    const jsonReport = JSON.parse(fs.readFileSync(reportPath, 'utf8'));
                    testResult.detailedResults = jsonReport;

                    // Update summary
                    if (jsonReport.numTotalTests) {
                        this.results.summary.totalTests += jsonReport.numTotalTests;
                        this.results.summary.passed += jsonReport.numPassedTests;
                        this.results.summary.failed += jsonReport.numFailedTests;
                        this.results.summary.skipped += jsonReport.numPendingTests;
                    }
                } catch (e) {
                    console.warn(`Warning: Could not parse JSON report: ${e.message}`);
                }
            }

            this.results.testResults[categoryName] = testResult;
            this.results.testRunInfo.testCategories.push(categoryName);

            console.log(`✅ ${categoryName} PASSED (${duration.toFixed(2)}s)`);
            return true;

        } catch (error) {
            const endTime = Date.now();
            const duration = (endTime - startTime) / 1000;

            const testResult = {
                category: categoryName,
                description: description,
                duration: duration,
                status: 'FAILED',
                error: error.message,
                output: error.stdout || '',
                stderr: error.stderr || ''
            };

            this.results.testResults[categoryName] = testResult;

            console.log(`❌ ${categoryName} FAILED (${duration.toFixed(2)}s)`);
            console.log(`Error: ${error.message}`);
            return false;
        }
    }

    analyzeSecurityIssues() {
        const securityIssues = [];
        const recommendations = [];

        for (const [category, result] of Object.entries(this.results.testResults)) {
            if (result.status === 'FAILED') {
                if (category.includes('security') || category.includes('auth')) {
                    securityIssues.push({
                        category: 'Mobile Security',
                        severity: 'HIGH',
                        issue: `Security tests failed in ${category}`,
                        description: 'Mobile security vulnerabilities detected'
                    });
                    recommendations.push({
                        category: 'Mobile Security',
                        priority: 'HIGH',
                        recommendation: 'Review and fix mobile security issues immediately'
                    });
                }

                if (category.includes('privacy') || category.includes('data')) {
                    securityIssues.push({
                        category: 'Data Privacy',
                        severity: 'HIGH',
                        issue: `Privacy tests failed in ${category}`,
                        description: 'Mobile data privacy issues detected'
                    });
                    recommendations.push({
                        category: 'Data Privacy',
                        priority: 'HIGH',
                        recommendation: 'Address mobile data privacy issues'
                    });
                }
            }
        }

        // Check for security-specific issues
        this.checkForCommonMobileSecurityIssues(securityIssues, recommendations);

        this.results.securityIssues = securityIssues;
        this.results.recommendations = recommendations;
    }

    checkForCommonMobileSecurityIssues(securityIssues, recommendations) {
        // Check package.json for known vulnerable dependencies
        try {
            const packageJson = JSON.parse(fs.readFileSync('package.json', 'utf8'));

            // Check for potentially insecure dependencies
            const insecureDeps = [
                'react-native-webview', // If using old versions
                'react-native-camera',  // If using deprecated versions
            ];

            const allDeps = { ...packageJson.dependencies, ...packageJson.devDependencies };

            for (const dep of insecureDeps) {
                if (allDeps[dep]) {
                    securityIssues.push({
                        category: 'Dependencies',
                        severity: 'MEDIUM',
                        issue: `Potentially insecure dependency: ${dep}`,
                        description: 'Review dependency for security vulnerabilities'
                    });
                }
            }
        } catch (e) {
            console.warn('Could not analyze package.json for security issues');
        }

        // Check for hardcoded secrets in source files
        this.checkForHardcodedSecrets(securityIssues, recommendations);
    }

    checkForHardcodedSecrets(securityIssues, recommendations) {
        const secretPatterns = [
            /api[_-]?key[_-]?=.{10,}/i,
            /secret[_-]?key[_-]?=.{10,}/i,
            /password[_-]?=.{5,}/i,
            /token[_-]?=.{10,}/i,
            /auth[_-]?key[_-]?=.{10,}/i
        ];

        const sourceFiles = this.getSourceFiles('src');

        for (const file of sourceFiles) {
            try {
                const content = fs.readFileSync(file, 'utf8');

                for (const pattern of secretPatterns) {
                    if (pattern.test(content)) {
                        securityIssues.push({
                            category: 'Hardcoded Secrets',
                            severity: 'HIGH',
                            issue: `Potential hardcoded secret in ${file}`,
                            description: 'Hardcoded secrets detected in source code'
                        });

                        recommendations.push({
                            category: 'Secrets Management',
                            priority: 'HIGH',
                            recommendation: 'Remove hardcoded secrets and use secure configuration'
                        });
                        break; // Only report once per file
                    }
                }
            } catch (e) {
                // Skip files that can't be read
            }
        }
    }

    getSourceFiles(dir, files = []) {
        const items = fs.readdirSync(dir);

        for (const item of items) {
            const fullPath = path.join(dir, item);
            const stat = fs.statSync(fullPath);

            if (stat.isDirectory() && !item.startsWith('.') && item !== 'node_modules') {
                this.getSourceFiles(fullPath, files);
            } else if (stat.isFile() && (item.endsWith('.js') || item.endsWith('.ts') || item.endsWith('.tsx'))) {
                files.push(fullPath);
            }
        }

        return files;
    }

    generateReport(outputFile = 'mobile-security-report.json') {
        this.analyzeSecurityIssues();

        // Calculate metrics
        const { totalTests, passed } = this.results.summary;
        if (totalTests > 0) {
            this.results.summary.passRate = Math.round((passed / totalTests) * 100 * 100) / 100;
        } else {
            this.results.summary.passRate = 0;
        }

        // Security score
        const securityScore = Math.max(0, 100 - (this.results.securityIssues.length * 10));
        this.results.summary.securityScore = securityScore;

        // Write report
        fs.writeFileSync(outputFile, JSON.stringify(this.results, null, 2));
        console.log(`\n📊 Mobile security test report generated: ${outputFile}`);

        return this.results;
    }

    printSummary() {
        console.log(`\n${'='.repeat(60)}`);
        console.log('MOBILE SECURITY TEST SUMMARY');
        console.log(`${'='.repeat(60)}`);

        const { summary } = this.results;

        console.log(`Total Tests: ${summary.totalTests}`);
        console.log(`Passed: ${summary.passed} ✅`);
        console.log(`Failed: ${summary.failed} ❌`);
        console.log(`Skipped: ${summary.skipped} ⏭️`);
        console.log(`Pass Rate: ${summary.passRate || 0}%`);
        console.log(`Security Score: ${summary.securityScore || 0}/100`);

        // Print security issues
        if (this.results.securityIssues.length > 0) {
            console.log(`\n🚨 SECURITY ISSUES DETECTED (${this.results.securityIssues.length})`);
            for (const issue of this.results.securityIssues) {
                const severityEmoji = {
                    'CRITICAL': '🔴',
                    'HIGH': '🟠',
                    'MEDIUM': '🟡',
                    'LOW': '🟢'
                }[issue.severity] || '⚪';

                console.log(`  ${severityEmoji} ${issue.severity}: ${issue.issue}`);
            }
        }

        // Print recommendations
        if (this.results.recommendations.length > 0) {
            console.log(`\n💡 RECOMMENDATIONS (${this.results.recommendations.length})`);
            for (const rec of this.results.recommendations) {
                const priorityEmoji = {
                    'CRITICAL': '🔴',
                    'HIGH': '🟠',
                    'MEDIUM': '🟡',
                    'LOW': '🟢'
                }[rec.priority] || '⚪';

                console.log(`  ${priorityEmoji} ${rec.priority}: ${rec.recommendation}`);
            }
        }

        console.log(`\n${'='.repeat(60)}`);
    }

    async runSecurityAudit() {
        console.log('\n🔍 Running npm security audit...');

        try {
            const auditResult = execSync('npm audit --json', { encoding: 'utf8' });
            const audit = JSON.parse(auditResult);

            if (audit.vulnerabilities && Object.keys(audit.vulnerabilities).length > 0) {
                console.log(`⚠️  Found ${Object.keys(audit.vulnerabilities).length} vulnerabilities`);

                for (const [pkg, vuln] of Object.entries(audit.vulnerabilities)) {
                    if (vuln.severity === 'high' || vuln.severity === 'critical') {
                        this.results.securityIssues.push({
                            category: 'Dependencies',
                            severity: vuln.severity.toUpperCase(),
                            issue: `Vulnerable dependency: ${pkg}`,
                            description: vuln.title || 'Security vulnerability in dependency'
                        });
                    }
                }
            } else {
                console.log('✅ No vulnerabilities found in dependencies');
            }
        } catch (error) {
            console.log('⚠️  Could not run security audit:', error.message);
        }
    }
}

async function main() {
    const args = process.argv.slice(2);
    const category = args.find(arg => arg.startsWith('--category='))?.split('=')[1];
    const output = args.find(arg => arg.startsWith('--output='))?.split('=')[1] || 'mobile-security-report.json';
    const noReport = args.includes('--no-report');

    // Create test reports directory
    if (!fs.existsSync('test-reports')) {
        fs.mkdirSync('test-reports');
    }

    const runner = new MobileSecurityTestRunner();

    // Define test categories
    const testCategories = [
        {
            name: 'authentication_security',
            pattern: '__tests__/**/auth/*.test.{js,ts,tsx}',
            description: 'Authentication and authorization security tests'
        },
        {
            name: 'data_privacy',
            pattern: '__tests__/**/privacy/*.test.{js,ts,tsx}',
            description: 'Data privacy and protection tests'
        },
        {
            name: 'network_security',
            pattern: '__tests__/**/services/*security*.test.{js,ts,tsx}',
            description: 'Network communication security tests'
        },
        {
            name: 'storage_security',
            pattern: '__tests__/**/storage/*.test.{js,ts,tsx}',
            description: 'Local storage security tests'
        },
        {
            name: 'e2e_security_workflows',
            pattern: '__tests__/e2e/*security*.test.{js,ts,tsx}',
            description: 'End-to-end security workflow tests'
        },
        {
            name: 'complete_user_workflows',
            pattern: '__tests__/e2e/complete-user-workflows.test.tsx',
            description: 'Complete user workflow security tests'
        }
    ];

    console.log('🔒 Starting Mobile Security Testing');
    console.log(`Timestamp: ${new Date().toISOString()}`);

    // Run security audit first
    await runner.runSecurityAudit();

    // Run specific category or all categories
    if (category) {
        const testCategory = testCategories.find(cat => cat.name === category);
        if (testCategory) {
            await runner.runTestCategory(testCategory.name, testCategory.pattern, testCategory.description);
        } else {
            console.log(`❌ Category '${category}' not found`);
            console.log('Available categories:');
            for (const cat of testCategories) {
                console.log(`  - ${cat.name}: ${cat.description}`);
            }
            process.exit(1);
        }
    } else {
        // Run all categories
        for (const testCategory of testCategories) {
            await runner.runTestCategory(testCategory.name, testCategory.pattern, testCategory.description);
        }
    }

    // Generate report
    if (!noReport) {
        runner.generateReport(output);
    }

    // Print summary
    runner.printSummary();

    // Exit with appropriate code
    if (runner.results.summary.failed > 0 || runner.results.securityIssues.length > 0) {
        console.log('\n❌ Mobile security tests failed or security issues detected!');
        process.exit(1);
    } else {
        console.log('\n✅ All mobile security tests passed!');
        process.exit(0);
    }
}

if (require.main === module) {
    main().catch(error => {
        console.error('Error running mobile security tests:', error);
        process.exit(1);
    });
}

module.exports = { MobileSecurityTestRunner };